File: src/api/schemas.py
from __future__ import annotations

from datetime import date
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class BacktestRunRequest(BaseModel):
    tickers: List[str] = Field(..., min_length=1)
    start_date: date
    end_date: date
    timeframe: str = "5m"
    initial_capital: float = 10000.0
    use_tws_data: bool = True


class BacktestRunResponse(BaseModel):
    job_id: str


class BacktestStatusResponse(BaseModel):
    job_id: str
    status: str
    error: Optional[str] = None
    started_at: Optional[str] = None
    finished_at: Optional[str] = None


class BacktestResultResponse(BaseModel):
    job_id: str
    status: str
    result: Optional[Dict[str, Any]] = None
    error: Optional[str] = None


class WatchlistUpdateRequest(BaseModel):
    symbols: List[str] = Field(default_factory=list)
    groups: Optional[List["WatchlistGroup"]] = None
    feed: Optional["WatchlistFeed"] = None


class WatchlistChangeRequest(BaseModel):
    symbol: str


class WatchlistItem(BaseModel):
    symbol: str = Field(..., min_length=1)
    exchange: str = ""
    name: str = ""
    enabled: bool = True
    instrument_id: Optional[str] = None


class WatchlistGroup(BaseModel):
    id: str = Field(..., min_length=1)
    name: str = Field(..., min_length=1)
    source: str = "manual"
    items: List[WatchlistItem] = Field(default_factory=list)


class WatchlistFeed(BaseModel):
    provider: str = Field(..., min_length=1)
    url: str = Field(..., min_length=1)
    title: Optional[str] = None
    external_id: Optional[str] = None
    last_refreshed_at: Optional[str] = None


class TradingViewWatchlistImportRequest(BaseModel):
    url: str = Field(..., min_length=1)


class WatchlistResponse(BaseModel):
    symbols: List[str] = Field(default_factory=list)
    groups: List[WatchlistGroup] = Field(default_factory=list)
    feed: Optional[WatchlistFeed] = None
    updated_at: Optional[str] = None


class CockpitStrategySummary(BaseModel):
    id: str = Field(..., min_length=1)
    name: str = Field(..., min_length=1)
    rule_count: int = 0
    enabled_rule_count: int = 0
    source: str = "library"


class CockpitStrategySlot(BaseModel):
    id: str = Field(..., min_length=1)
    label: str = Field(..., min_length=1)
    strategy_id: Optional[str] = None
    enabled: bool = True


class CockpitWorkspace(BaseModel):
    id: str = Field(..., min_length=1)
    name: str = Field(..., min_length=1)
    kind: str = "long"
    enabled: bool = True
    strategy_slots: List[CockpitStrategySlot] = Field(default_factory=list)


class CockpitStateUpdateRequest(BaseModel):
    global_enabled: bool = True
    active_workspace_id: Optional[str] = None
    workspaces: List[CockpitWorkspace] = Field(default_factory=list)


class CockpitStateResponse(BaseModel):
    global_enabled: bool = True
    active_workspace_id: Optional[str] = None
    workspaces: List[CockpitWorkspace] = Field(default_factory=list)
    strategy_library: List[CockpitStrategySummary] = Field(default_factory=list)
    feed: Optional[WatchlistFeed] = None
    updated_at: Optional[str] = None


class StrategyLibraryEntry(BaseModel):
    id: str = Field(..., min_length=1)
    name: str = Field(..., min_length=1)
    rule_count: int = 0
    enabled_rule_count: int = 0
    updated_at: Optional[str] = None


class StrategyLibrarySaveRequest(BaseModel):
    strategy: Dict[str, Any]
    name: Optional[str] = None


class ConfigUpdateRequest(BaseModel):
    updates: Dict[str, Dict[str, Any]]


class TWSConnectionRequest(BaseModel):
    host: Optional[str] = None
    port: Optional[int] = None
    client_id: Optional[int] = None


class BotStateIngestRequest(BaseModel):
    state: Dict[str, Any]


class BotCommand(BaseModel):
    id: int
    command: str
    payload: Optional[Dict[str, Any]] = None
    created_at: Optional[str] = None


class BotCommandResponse(BaseModel):
    commands: List[BotCommand]


class DiagnosticsStartup(BaseModel):
    environment: str
    trading_mode: str
    ib_host: str
    ib_port: int
    client_id: int
    account: str
    log_level: str
    log_file: str
    watchlist_path: str
    strategy_path: str
    symbol_cache_path: str


class DiagnosticsRuntime(BaseModel):
    runner_active: bool = False
    last_runtime_reload_at: Optional[str] = None
    last_runtime_reload_reason: Optional[str] = None
    last_disconnect_at: Optional[str] = None
    last_disconnect_reason: Optional[str] = None


class DiagnosticsSymbols(BaseModel):
    source: Optional[str] = None
    last_checked_at: Optional[str] = None
    last_warning: Optional[str] = None


class DiagnosticsResponse(BaseModel):
    startup: DiagnosticsStartup
    runtime: DiagnosticsRuntime
    symbols: DiagnosticsSymbols


class NotificationTestRequest(BaseModel):
    message: str = Field(..., min_length=1)
    channel: Optional[str] = Field(None, description="Optional channel: telegram or discord")


class NotificationTestResponse(BaseModel):
    status: str


class NotificationEvent(BaseModel):
    id: str
    message: str
    level: str = "info"
    channel: Optional[str] = None
    created_at: str


class NotificationListResponse(BaseModel):
    events: List[NotificationEvent]
    total: int
    page: int
    page_size: int

File: backend/models/schemas.go
package models

// This package mirrors the FastAPI Pydantic schemas from tws_traderbot/src/api/schemas.py

type WatchlistItem struct {
	Symbol       string `json:"symbol"`
	Exchange     string `json:"exchange"`
	Name         string `json:"name"`
	Enabled      bool   `json:"enabled"`
	InstrumentID string `json:"instrument_id,omitempty"`
}

type WatchlistGroup struct {
	ID     string          `json:"id"`
	Name   string          `json:"name"`
	Source string          `json:"source"`
	Items  []WatchlistItem `json:"items"`
}

type WatchlistFeed struct {
	Provider        string `json:"provider"`
	URL             string `json:"url"`
	Title           string `json:"title,omitempty"`
	ExternalID      string `json:"external_id,omitempty"`
	LastRefreshedAt string `json:"last_refreshed_at,omitempty"`
}

type WatchlistResponse struct {
	Symbols   []string         `json:"symbols"`
	Groups    []WatchlistGroup `json:"groups"`
	Feed      *WatchlistFeed   `json:"feed,omitempty"`
	UpdatedAt string           `json:"updated_at,omitempty"`
}

type CockpitStrategySummary struct {
	ID               string `json:"id"`
	Name             string `json:"name"`
	RuleCount        int    `json:"rule_count"`
	EnabledRuleCount int    `json:"enabled_rule_count"`
	Source           string `json:"source"`
}

type CockpitStrategySlot struct {
	ID         string `json:"id"`
	Label      string `json:"label"`
	StrategyID string `json:"strategy_id,omitempty"`
	Enabled    bool   `json:"enabled"`
}

type CockpitWorkspace struct {
	ID            string                `json:"id"`
	Name          string                `json:"name"`
	Kind          string                `json:"kind"`
	Enabled       bool                  `json:"enabled"`
	StrategySlots []CockpitStrategySlot `json:"strategy_slots"`
}

type CockpitStateResponse struct {
	GlobalEnabled     bool                     `json:"global_enabled"`
	ActiveWorkspaceID string                   `json:"active_workspace_id,omitempty"`
	Workspaces        []CockpitWorkspace       `json:"workspaces"`
	StrategyLibrary   []CockpitStrategySummary `json:"strategy_library"`
	Feed              *WatchlistFeed           `json:"feed,omitempty"`
	UpdatedAt         string                   `json:"updated_at,omitempty"`
}
